hw/ip/otbn/dv/otbnsim/sim/stats.py

```diff
@@ -37,33 +37,40 @@ def record_stall(self) -> None:
 
     def record_insn(self,
                     insn: OTBNInsn,
-                    state_before_commit: OTBNState) -> None:
-        ''' Record the execution of an instruction. '''
-        pc = state_before_commit.pc
+                    state_bc: OTBNState) -> None:
+        '''Record the execution of an instruction.
+
+        insn is the currently executed instruction. state_bc is the state of
+        OTBN the instruction is committed.
+
+        '''
+        pc = state_bc.pc
+
+        is_jump = isinstance(insn, JAL) or isinstance(insn, JALR)
 
         # Instruction histogram
         self.insn_histo[insn.insn.mnemonic] += 1
 
         # Function calls
         # - Direct function calls: jal x1, <offset>
         # - Indirect function calls: jalr x1, <grs1>, 0
-        if (isinstance(insn, JAL) or isinstance(insn, JALR)) and insn.grd == 1:
-            call_stack = state_before_commit.peek_call_stack()
+        if is_jump and insn.grd == 1:  # type: ignore
```

Comment:
That's sad. Presumably mypy can't follow the types through setting `is_jump`. Is there any reason to change this from what was there before, which didn't need the waiver?

Reply:
is_jump is going to be used again in the basic block stats PR (together with a is_branch) and I want to avoid duplicating this code. Unfortunately mypy got priority 2 behind "reducing code duplication".

I tried a couple options but I couldn't really find a readable workaround that wouldn't require effectively duplicating the JAL/JALR check.

Reply:
Oh, I see. Boo :-(